Support declarativeNetRequest.getMatchedRules

b-weinstein opened the pull request:
#### e270f83dd685dc0262a6ab8a12bedf26a0560049
<pre>
Support declarativeNetRequest.getMatchedRules
<a href="https://bugs.webkit.org/show_bug.cgi?id=265917">https://bugs.webkit.org/show_bug.cgi?id=265917</a>
<a href="https://rdar.apple.com/118940129">rdar://118940129</a>

Reviewed by Timothy Hatcher.

This patch keeps track of the content rule list actions that each extension has performed and
exposes them through declarativeNetRequest.getMatchedRules().

These can be filtered on tabId and timeStamp, and are limited to only vending information about
URLs that the extension has access to.

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Shared/Extensions/WebExtensionMatchedRuleParameters.h: Added.
* Source/WebKit/Shared/Extensions/WebExtensionMatchedRuleParameters.serialization.in: Added.
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIDeclarativeNetRequestCocoa.mm:
(WebKit::WebExtensionContext::declarativeNetRequestGetMatchedRules): Iterate over the matched rules and
filter them. Also check to make sure we have permission to access the URLs before vending them.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::unload): Clear the matched rules.
(WebKit::WebExtensionContext::handleContentRuleListNotificationForTab): Increment the blocked resource count
and add the blocked resource to the list of matched rules.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
(WebKit::WebExtensionController::handleContentRuleListNotification):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
(WebKit::WebExtensionContext::matchedRules): Keep track of the matched rules.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIDeclarativeNetRequestCocoa.mm:
(WebKit::extensionHasPermission): Add a helper method.
(WebKit::toWebAPI): Ditto.
(WebKit::WebExtensionAPIDeclarativeNetRequest::getMatchedRules):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIDeclarativeNetRequest.mm:
(TestWebKitAPI::TEST): Add a test that getMatchedRules returns a match that the frame was blocked.

Canonical link: <a href="https://commits.webkit.org/271593@main">https://commits.webkit.org/271593@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b6984457544faa9d47bd03111229591186c4a3b3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/28928 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/7574 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/30284 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/31543 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/26366 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/29501 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/9700 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/4917 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/26397 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/29200 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/6263 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/24831 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/5459 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/5580 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/25855 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/32880 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/26455 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/26288 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/31833 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/5561 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/3734 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/29609 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/7203 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6911 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/6046 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/6082 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIDeclarativeNetRequestCocoa.mm:
```diff
@@ -222,6 +222,46 @@
 
+void WebExtensionContext::declarativeNetRequestGetMatchedRules(std::optional<WebExtensionTabIdentifier> tabIdentifier, std::optional<double> minTimeStamp, CompletionHandler<void(std::optional<Vector<WebExtensionMatchedRuleParameters>> matchedRules, std::optional<String>)>&& completionHandler)
+{
+    RefPtr tab = tabIdentifier ? getTab(tabIdentifier.value()) : nil;
```
`nullptr` instead of `nil`.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIDeclarativeNetRequestCocoa.mm:
```diff
@@ -222,6 +222,46 @@
+    RefPtr tab = tabIdentifier ? getTab(tabIdentifier.value()) : nil;
+
+    if (tabIdentifier && !tab) {
+        completionHandler(std::nullopt, toErrorString(@"declarativeNetRequest.getMatchedRules()", @"tabId", @"The specified tab does not exist."));
```
`toErrorString(@"declarativeNetRequest.getMatchedRules()", nil, @"tab not found")` to match other API errors.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIDeclarativeNetRequestCocoa.mm:
```diff
@@ -222,6 +222,46 @@
+        ASSERT(hasPermission(_WKWebExtensionPermissionActiveTab));
+
+        if (!hasPermission(_WKWebExtensionPermissionTabs, tab.get())) {
+            completionHandler(std::nullopt, toErrorString(@"declarativeNetRequest.getMatchedRules()", nil, @"The 'activeTab' permission has not been granted by the user for the specified tab."));
```
Should break out `@"declarativeNetRequest.getMatchedRules()"` into `static NSString * const apiName` since you use it twice here.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIDeclarativeNetRequestCocoa.mm:
```diff
@@ -222,6 +222,46 @@
+    for (auto matchedRule : matchedRules()) {
+        // FIXME: https://bugs.webkit.org/show_bug.cgi?id=262714 - we should be checking for permission if we don't have access to this URL.
+        if (!hasPermission(matchedRule.url))
+            continue;
```
You are checking permission, but this FIXME is about _requesting_ permission.

b-weinstein replied:
Yeah - I'll update the wording of the FIXME.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -3010,6 +3012,24 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
+    if (!hasPermission(_WKWebExtensionPermissionDeclarativeNetRequestFeedback) && !(hasPermission(_WKWebExtensionPermissionDeclarativeNetRequest) && hasPermission(_WKWebExtensionPermissionActiveTab)))
+        return;
+
+    // FIXME: Set a timer to purge old matched rules.
```
See `WebExtensionAlarm.cpp` for a simple-ish timer example.

b-weinstein replied:
I'll put this into the next patch, and add a bug for the FIXME in the meantime.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -3010,6 +3012,24 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
+    static constexpr double oneSecondInMilliseconds = 1000;
+    m_matchedRules.append({
+        url,
+        NSDate.now.timeIntervalSince1970 * oneSecondInMilliseconds,
```
`WallTime::now()`

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionMatchedRuleParameters.h:
```diff
@@ -0,0 +1,43 @@
+
+struct WebExtensionMatchedRuleParameters {
+    URL url;
+    double timeStamp;
```
This would be better as `WallTime`. It will be easier to work with in other parts of this PR.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIDeclarativeNetRequestCocoa.mm:
```diff
@@ -222,6 +222,46 @@
+    }
+
+    static constexpr double oneSecondInMilliseconds = 1000;
+    NSDate *minTimeStampDate = minTimeStamp ? [NSDate dateWithTimeIntervalSince1970:minTimeStamp.value() / oneSecondInMilliseconds] : nil;
```
If you use `WallTime` for `minTimeStamp`:

`WallTime minTime = minTimeStamp ? minTimeStamp.value() : WallTime::nan();`

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIDeclarativeNetRequestCocoa.mm:
```diff
@@ -222,6 +222,46 @@
+
+        NSDate *timeStampDate = [NSDate dateWithTimeIntervalSince1970:matchedRule.timeStamp / oneSecondInMilliseconds];
+        if (minTimeStampDate && [timeStampDate compare:minTimeStampDate] != NSOrderedDescending)
+            continue;
```
Assuming you use `WallTime` for `matchedRule.timeStamp`:

```
if (minTime != WallTime::nan() && matchedRule.timeStamp <= minTime)
    continue;
```

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIDeclarativeNetRequestCocoa.mm:
```diff
@@ -107,9 +110,75 @@
+
+        [matchedRuleArray addObject:@{
+            @"request": @{ @"url": urlString },
+            @"timeStamp": @(timeStamp),
```
With `WallTime`:

`@(floor(timeStamp.secondsSinceEpoch().milliseconds()))`

xeenon replied:
While I'm here `WebExtensionContextProxy::dispatchWebNavigationEvent` should do `floor()` too.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIDeclarativeNetRequestCocoa.mm:
```diff
@@ -107,9 +110,75 @@
+    }
+
+    NSNumber *minTimeStamp = objectForKey<NSNumber>(filter, getMatchedRulesMinTimeStampKey);
+    std::optional<double> optionalTimeStamp = std::nullopt;
```
It defaults to `std::nullopt`, so you can omit that part.

xeenon replied:
`std::optional<WallTime>` instead of `double`

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIDeclarativeNetRequestCocoa.mm:
```diff
@@ -107,9 +110,75 @@
+    NSNumber *minTimeStamp = objectForKey<NSNumber>(filter, getMatchedRulesMinTimeStampKey);
+    std::optional<double> optionalTimeStamp = std::nullopt;
+    if (minTimeStamp)
+        optionalTimeStamp = minTimeStamp.doubleValue;
```
With `WallTime`:

`optionalTimeStamp = WallTime::fromRawSeconds(Seconds::fromMilliseconds(minTimeStamp.doubleValue).value())`

xeenon commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIDeclarativeNetRequest.mm:
```diff
@@ -436,6 +436,65 @@
     EXPECT_NS_EQUAL(action.badgeText, @"3");
 }
 
+TEST(WKWebExtensionAPIDeclarativeNetRequest, GetMatchedRules)
```
So nice!

xeenon commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIDeclarativeNetRequest.mm:
```diff
@@ -436,6 +436,65 @@
+    auto *urlRequest = server.requestWithLocalhost();
+    NSURL *requestURL = urlRequest.URL;
+    [manager.get().context setPermissionStatus:_WKWebExtensionContextPermissionStatusGrantedExplicitly forURL:requestURL];
+    [manager.get().context setPermissionStatus:_WKWebExtensionContextPermissionStatusGrantedExplicitly forURL:[requestURL URLByAppendingPathComponent:@"frame.html"]];
```
`server.requestWithLocalhost("/frame.html"_s)` will give you a request for that path. But what you have is fine too.

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIDeclarativeNetRequestCocoa.mm:
```diff
@@ -107,9 +110,75 @@
+
+void WebExtensionAPIDeclarativeNetRequest::getMatchedRules(NSDictionary *filter, Ref<WebExtensionCallbackHandler>&& callback, NSString **outExceptionString)
+{
+    BOOL hasFeedbackPermission = extensionHasPermission(extensionContext(), _WKWebExtensionPermissionDeclarativeNetRequestFeedback);
```
`bool`

b-weinstein committed:
Support declarativeNetRequest.getMatchedRules
https://bugs.webkit.org/show_bug.cgi?id=265917
rdar://118940129

Reviewed by Timothy Hatcher.

This patch keeps track of the content rule list actions that each extension has performed and
exposes them through declarativeNetRequest.getMatchedRules().

These can be filtered on tabId and timeStamp, and are limited to only vending information about
URLs that the extension has access to.

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Shared/Extensions/WebExtensionMatchedRuleParameters.h: Added.
* Source/WebKit/Shared/Extensions/WebExtensionMatchedRuleParameters.serialization.in: Added.
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIDeclarativeNetRequestCocoa.mm:
(WebKit::WebExtensionContext::declarativeNetRequestGetMatchedRules): Iterate over the matched rules and
filter them. Also check to make sure we have permission to access the URLs before vending them.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::unload): Clear the matched rules.
(WebKit::WebExtensionContext::handleContentRuleListNotificationForTab): Increment the blocked resource count
and add the blocked resource to the list of matched rules.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
(WebKit::WebExtensionController::handleContentRuleListNotification):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
(WebKit::WebExtensionContext::matchedRules): Keep track of the matched rules.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIDeclarativeNetRequestCocoa.mm:
(WebKit::extensionHasPermission): Add a helper method.
(WebKit::toWebAPI): Ditto.
(WebKit::WebExtensionAPIDeclarativeNetRequest::getMatchedRules):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIDeclarativeNetRequest.mm:
(TestWebKitAPI::TEST): Add a test that getMatchedRules returns a match that the frame was blocked.

Canonical link: https://commits.webkit.org/271593@main
